### ISSUE TITLE
Can't read footer in an object data file

### ISSUE THREAD

**Issue by mmmikemon** [ccd-utexas/tsphot#108]
Using the pipeline branch of tsphot I have an issue reading the footer in an object data file. The file in question is on lonestar/corral and is
/corral-repl/utexas/White_Dwarf_Archive/ProEM_1024B/20141121/SDSSJ0651+2844/SDSSJ0651+2844.spe

Essentially, read_spe returns an empty footer, which leads to errors down the line.

**Comment by keatonb** [comment]
Seconded.  Following these directions http://nbviewer.ipython.org/gist/ccd-admin/e3f5a7ce74f69a1f8ae5 does not work for me either.  When I run the 4th cell, I do not get the warning "INFO: XML footer metadata is empty for...".  This is because hasattr(spe, 'footer_metadata') is True in this case, but getattr(spe, 'footer_metadata') == ''.

> Adding the check that getattr(_) != '', I get the result:
> Trigger response = 
> Absolute timestamp = 2015-03-13 23:48:21
> Ticks per second = 1000000.0

> WARNING: No XML footer metadata.
> Unknown trigger response.
> Using file creation time as absolute timestamp.
> Assuming 1E6 ticks per seconds.

Still, I'd like to see confirmation that this is a safe assumption.  How does file creation time relate to absolute timestamp.  In this example, the first frame saved is #7.  How are we sure this is properly taken into account?

**Comment by keatonb** [comment]
After a double take, this is emphatically NOT an appropriate way to recover the timestamp, since it returned *today's* date!

**Comment by stharrold** [comment]
Hence all my disclaimers ;)

**Comment by keatonb** [comment]
Sam, I'm not sure which disclaimers you are referring to.  That you have a "Extract timestamps from any SPE file (corrupted or not)" script in the documentation for tsphot led me to believe that this lack of footer info is a nonissue.

Please appreciate that these data are for one of our group's most important multi-year observing projects and are rendered completely useless if we cannot trust the timestamps.  I find your use of emoticons highly inappropriate.

This highlights the fact that a thorough timestamp verification pipeline is critically needed, and should have been the top priority from the onset of using SPE files.  Obviously this needs to be the priority now, with ample test cases.

**Comment by stharrold** [comment]
Doing what I can Keaton. You're welcome that we can observe at all.
#readthedocs

**Comment by keatonb** [comment]
It appears that the timestamps.csv file in that directory *might* be reliable, as it seems to agree with the manual log.

Any idea what the two timestamps that LightField puts into the .fits files that it extracts from the SPE file mean?  Are they both related to when the extraction was done?  I can't seem to make sense of the fits header numbers for the run in question.

**Comment by stharrold** [comment]
I'm glad that you're observing to validate timings. To help your validations, you may want to export the SPE per-frame metadata to CSV using the built-in LightField utility (see ["End of Night" section of Observing_Guide.gdoc](https://docs.google.com/document/d/1LLg7ABohlfWF3TCUaFNpVRdKJGi9brlw6kLK4DFjf50/edit#heading=h.fxkxnsyc6t5) and Ch 9 of the [LightField User's Manual](ftp://ftp.princetoninstruments.com/Public/Manuals/Princeton%20Instruments/LightField%20User's%20Manual%20Ver.%204.5.pdf))

Here's a gist that shows how I inspected the data you took last night on Lonestar (tsphot is on Lonestar):
https://gist.github.com/stharrold/e5637275504277383ba3

Here's the relevant data from the gist (see the interpretation below):
- from G117-B15A-Frame-0001.fits (inspected using `head`):  
DATE    = '2015-03-14T12:12:42' / file creation date (YYYY-MM-DDThh:mm:ss UT)  
DATE-OBS= '2015-03-14T08:36:31' / Date of data acquisition 
- from the SPE XML footer (inspected using `tail` and `read_spe`):  
```<TimeStamp event="ExposureStarted" type="Int64" bitDepth="64" resolution="1000000" absoluteTime="2015-03-14T05:31:40.9319956Z" /><TimeStamp event="ExposureEnded" type="Int64" bitDepth="64" resolution="1000000" absoluteTime="2015-03-14T05:31:40.9319956Z" /> ```
```<FileInformation creator="ARGOS-DEV\\admin" created="2015-03-14T08:36:31.4082176Z" lastModified="2015-03-14T08:36:31.4082176Z" />```  

- Interpretation:
  - The ProEM was made for high-speed experiments with short durations (i.e. the entire SPE file was created in less than 1 second). The FITS files created by Princeton Instruments do not necessarily reflect that.
  - The "DATE" keyword from G117-B15A-Frame-0001.fits appears to be the datetime stamp of when you clicked "Export to Fits" on LightField.
  - The "DATE-OBS" keyword from G117-B15A-Frame-0001.fits appears to be the file creation time of the SPE file, i.e. when you clicked "Stop" (c.f. `FileInformation` tag from the SPE XML footer.)
  - You clicked "Run All" at 2015-03-14T05:31:40.9319956Z (c.f. `TimeStamp` tag from the SPE XML footer).

- Next steps:
  - I'll look at the problem data from 2014-11-21.

**Comment by stharrold** [comment]
Also for validating timings, consider adding the date and time stamp to the filename as suggested in ("default_experiment.lfe" section of Observing_Guide.gdoc)[https://docs.google.com/document/d/1LLg7ABohlfWF3TCUaFNpVRdKJGi9brlw6kLK4DFjf50/edit#heading=h.7qhuu04vde4p]

**Comment by stharrold** [comment]
LightField read SDSSJ0651+2844.spe correctly. Next step is to look at how `read_spe.py` extracts the XML footer.

![20150314t130000_lightfield_capture](https://cloud.githubusercontent.com/assets/4134823/6652806/d18656c0-ca49-11e4-870d-ba8755541e02.png)

**Comment by stharrold** [comment]
Example inspecting the problem data from 2014-11-21:
https://gist.github.com/stharrold/ce469bbe797118c2a591

**Comment by keatonb** [comment]
> Also for validating timings, consider adding the date and time stamp to the filename
Good idea.  This has generally been avoided since it causes headaches with IRAF, but hopefully we're moving to an all-python reduction pipeline soon that the headaches will subside soon.

> LightField read the footer metadata from SDSSJ0651+2844.spe correctly. Next step is to look at how read_spe.py extracts the XML footer.
Ahh... a sigh of relief.

**Comment by keatonb** [comment]
What are the JD timestamps based on in the Online Reduction's lightcurve.txt file?  Those times are from before the footer was supposed to be written out, correct?

**Comment by stharrold** [comment]
Yes, if you're acquiring data while reducing it, then the XML footer has not been written yet. Most of the cookbook recipes I've written use the file creation time as a substitute for the initial time stamp (`os.path.getctime`). So if you copy-paste a file or rename a file, that creation time will probably be different from when LightField actually first wrote the file.

**Comment by stharrold** [comment]
The IPython Notebook I used for troubleshooting `read_spe._load_footer_metadata` with the built-in python debugger `pdb`
http://nbviewer.ipython.org/gist/stharrold/32861b0ae60b166e2f03

Everything I did on the command line to fix this bug (mostly just how to work with git from the command line and how to run `py.test`):
https://gist.github.com/stharrold/25f7889f4fcf1111d6bd

**Comment by stharrold** [comment]
Summary: 
- The problem was that the file header set the byte offset for the start of the XML footer 4 bytes after the actual start of the XML footer. I the parsed the partial XML footer in a way that left an empty string.
- This problem only seems to occur when users click "Run All" then "Acquire" rather than "Run All", "Stop", then "Acquire".
- My fix was to start reading the XML footer 1KB before the byte offset of the start of the XML footer. I then trim the excess binary data from the XML string. If the XML string is not properly formatted, I print a warning to stderr then return the raw string. (This rare case has never been encountered, but in that event, you'll have to parse the string yourself.)
- This is an error from LightField and may be worth reporting if you want to send. However, it is very minor since it was easy to fix and LightField has no problem reading the XML even with the wrong byte offset (from the screenshot).
- See the links above for how to do what I did, and see this link for the actual code diff: Graphs > Network > click the node where `hotfix` was merged back into `master` https://github.com/ccd-utexas/tsphot/commit/9ed590cc8d14616bf6500871ade25f0c63aedb20
